### ISSUE TITLE
App Service IP Restrictions headers required even if not applicable

### ISSUE THREAD

**Issue by torivara** [hashicorp/terraform-provider-azurerm#12367]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform 0.15.5
Azurerm_provider 2.65.0

### Affected Resource(s)

* `azurerm_app_service.site_config.ip_restriction`

### Terraform Configuration Files

```hcl
resource "azurerm_app_service" "app_service" {
  name                = "appservice"
  location            = "norwayeast"
  resource_group_name = "app-resourcegroup"
  app_service_plan_id = var.app_service_plan_id

  site_config {
    ip_restriction              = {
      name                      = "cidr_restriction_001"
      ip_address                = "1.2.3.4/32"
      virtual_network_subnet_id = null
      service_tag               = null
      subnet_id                 = null
      priority                  = 100
      action                    = "Allow"

      headers = [{
        x_azure_fdid      = null
        x_fd_health_probe = null
        x_forwarded_for   = null
        x_forwarded_host  = null
      }]
    }
  }
}
```

### Expected Behaviour

The actual expected behaviour is that ip_restrictions.headers is an optional setting. According to the documentation the headers block should be optional, but it is not. https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/app_service#ip_restriction

First run, it should set ip_restrictions with headers to null. Subsequent runs should see that the headers are already nulled, and notice that no change is required.

### Actual Behaviour

If run without the headers block, plan step fails with the following:

Error: Incorrect attribute value type
│ 
│   on ../../modules/appserviceweb/main.tf line 28, in resource "azurerm_app_service" "app_service":
│   28:       ip_restriction              = concat(local.subnets, local.cidrs, local.service_tags)
│     ├────────────────
│     │ local.cidrs is tuple with 1 element
│     │ local.service_tags is empty tuple
│     │ local.subnets is tuple with 1 element
│ 
│ Inappropriate value for attribute "ip_restriction": element 0: attribute
│ "headers" is required.

If you run with a headers block defined, it will be re-deployed each run even if configuration is not changed.

### Steps to Reproduce

1. Deploy an app with ip_restrictions and no headers block.
2. The deployment fails as headers block is apparently required.
3. Add a headers block and apply the configuration. It applies successfully.
4. Attempt to re-apply the same configuration without config change, and the headers block is planned for change.

### References

Have not found any similar issues.

**Comment by panderso01** [comment]
also seeing service_tag is required when it should be optional

**Comment by torivara** [comment]
Same issue persists after upgrading to Terraform version 1.0.3 and azurerm-provider 2.69.0. 

Can anyone comment on how I can ignore changes on this as a temporary workaround? Because of some dynamic blocks in the app service module, I am struggling to ignore the headers block.

Error: Unsupported attribute
│ 
│   on ../../modules/appserviceweb/main.tf line 124, in resource "azurerm_app_service" "app_service":
│  124:       site_config[0].ip_restriction.headers
│ 
│ Can't access attributes on a list of objects. Did you mean to access
│ attribute "headers" for a specific element of the list, or across all
│ elements of the list?

**Comment by torivara** [comment]
After some trial and error, it seems you can put the code as below, and there is no need to ignore changes. Still, it seems to me that the headers-block should be optional, and assumed empty if not provided.

Please note that this is part of a modified version of [this](https://github.com/claranet/terraform-azurerm-app-service-web) module, and takes variables for most settings.

```
#Before (results in error)
cidrs = [for cidr in var.authorized_ips : {
    name                      = "ip_restriction_cidr_${join("", [1, index(var.authorized_ips, cidr)])}"
    ip_address                = cidr
    virtual_network_subnet_id = null
    service_tag               = null
    subnet_id                 = null
    priority                  = join("", [1, index(var.authorized_ips, cidr)])
    action                    = "Allow"

     headers = [{
       x_azure_fdid      = null
       x_fd_health_probe = null
       x_forwarded_for   = null
       x_forwarded_host  = null
     }]
  }]

#After (results in "no infrastructure changes")
cidrs = [for cidr in var.authorized_ips : {
    name                      = "ip_restriction_cidr_${join("", [1, index(var.authorized_ips, cidr)])}"
    ip_address                = cidr
    virtual_network_subnet_id = null
    service_tag               = null
    subnet_id                 = null
    priority                  = join("", [1, index(var.authorized_ips, cidr)])
    action                    = "Allow"

    headers = []
  }]
```

**Comment by p0onage** [comment]
Same issue here, had to declare  the below in ip_restrictions 
```
 headers = []
 service_tag               = null
```
ordering also mattered. 
azurerm :   version = "~>2.93.0"
Terraform :  v0.14.7

**Comment by katbyte** [comment]
Worth noting the new app service resources released in 3.0 make this property optional:

Linux: [registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/linux_web_app](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/linux_web_app)
Windows: [registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/windows_web_app](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/windows_web_app)

More information on the Preview can be found in the 3.0 overview guide: [registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/3.0-overview](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/3.0-overview)

as the old resources have been frozen and the new ones fix this I'm going to close this out.
